File: source/com/janilla/todomvc/TodoMVC.js
/*
 * MIT License
 *
 * Copyright (c) 2024 Diego Schivo
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import Filters from './Filters.js';
import Rendering from './Rendering.js';
import TodoCount from './TodoCount.js';
import TodoList from './TodoList.js';
import ToggleAll from './ToggleAll.js';

class TodoMVC {

	selector;

	todos = [];

	uniqueID = 1;

	todoList;

	todoCount;

	filters;

	filter;

	get count() {
		const t = this.todos.length;
		const c = this.todos.reduce((s, t) => t.completed ? s + 1 : s, 0);
		return {
			active: t - c,
			completed: c,
			total: t
		};
	}

	run = async () => {
		const r = new Rendering();
		this.selector().innerHTML = await r.render(this, 'TodoMVC');
		this.listen();
	}

	render = async key => {
		switch (key) {
			case 'toggleAll':
				this.toggleAll = new ToggleAll();
				this.toggleAll.selector = () => this.selector().querySelector('.toggle-all-container');
				return this.toggleAll;

			case 'todoList':
				this.todoList = new TodoList();
				this.todoList.selector = () => this.selector().querySelector('.todo-list');
				return this.todoList;

			case 'todoCount':
				this.todoCount = new TodoCount();
				this.todoCount.selector = () => this.selector().querySelector('.todo-count');
				return this.todoCount;

			case 'filters':
				this.filters = new Filters();
				this.filters.selector = () => this.selector().querySelector('.filters');
				return this.filters;
		}
	}

	listen = () => {
		addEventListener('hashchange', this.handleHashChange);

		const e = this.selector();
		e.querySelector('.new-todo').addEventListener('change', this.handleNewTodoChange);
		e.addEventListener('todotoggleall', this.handleTodoToggleAll);
		e.addEventListener('todotoggle', this.handleTodoToggle);
		e.addEventListener('tododestroy', this.handleTodoDestroy);
		e.addEventListener('todoedit', this.handleTodoEdit);
		e.querySelector('.clear-completed').addEventListener('click', this.handleClearCompletedClick);

		this.todoList.listen();
		this.todoCount.listen();
		this.filters.listen();
	}

	handleHashChange = () => {
		this.filter = location.hash.substring(2);
		this.toggleAll.refresh();
		this.todoList.refresh();
		this.todoCount.refresh();
		this.filters.refresh();
	}

	handleNewTodoChange = e => {
		const i = e.currentTarget;
		const t = i.value.trim();
		if (t === '')
			return;
		const u = {
			id: this.uniqueID++,
			title: t,
			completed: false
		}
		this.todos.push(u);
		i.value = '';
		this.toggleAll.refresh();
		this.todoList.refresh();
		this.todoCount.refresh();
		this.contentBlock();
	}

	handleTodoToggleAll = e => {
		const c = e.detail.completed;
		this.todos.forEach(u => u.completed = c);
		this.toggleAll.refresh();
		this.todoList.refresh();
		this.todoCount.refresh();
	}

	handleTodoToggle = e => {
		const t = e.detail.todo;
		t.completed = !t.completed;
		this.toggleAll.refresh();
		this.todoList.refresh();
		this.todoCount.refresh();
	}

	handleTodoDestroy = e => {
		const t = e.detail.todo;
		this.todos.splice(this.todos.indexOf(t), 1);
		this.toggleAll.refresh();
		this.todoList.refresh();
		this.todoCount.refresh();
		this.contentBlock();
	}

	handleTodoEdit = e => {
		const t = e.detail.todo;
		t.title = e.detail.title;
		this.todoList.refresh();
	}

	handleClearCompletedClick = e => {
		this.todos = this.todos.filter(t => !t.completed);
		this.toggleAll.refresh();
		this.todoList.refresh();
		this.contentBlock();
	}
	
	contentBlock = () => {
		const e = this.selector();
		const d = this.todos.length ? '' : 'none';
		['.main', '.footer'].forEach(s => e.querySelector(s).style.display = d);
	}
}

export default TodoMVC;

File: main/source/com/janilla/todomvc/TodoMvc.java
/*
 * MIT License
 *
 * Copyright (c) 2024-2026 Diego Schivo
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.janilla.todomvc;

import java.lang.reflect.Modifier;
import java.net.InetSocketAddress;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Properties;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Stream;

import javax.net.ssl.SSLContext;

import com.janilla.http.HttpHandler;
import com.janilla.http.HttpServer;
import com.janilla.ioc.DiFactory;
import com.janilla.java.Java;
import com.janilla.net.Net;
import com.janilla.web.ApplicationHandlerFactory;
import com.janilla.web.Handle;
import com.janilla.web.Invocable;
import com.janilla.web.NotFoundException;
import com.janilla.web.Render;
import com.janilla.web.RenderableFactory;

@Render(template = "index.html")
public class TodoMvc {

	public static final AtomicReference<TodoMvc> INSTANCE = new AtomicReference<>();

	public static void main(String[] args) {
		try {
			TodoMvc a;
			{
				var f = new DiFactory(Stream.of(TodoMvc.class.getPackageName(), "com.janilla.web")
						.flatMap(x -> Java.getPackageClasses(x).stream()).toList(), INSTANCE::get);
				a = f.create(TodoMvc.class, Java.hashMap("diFactory", f, "configurationFile", args.length > 0 ? Path
						.of(args[0].startsWith("~") ? System.getProperty("user.home") + args[0].substring(1) : args[0])
						: null));
			}

			HttpServer s;
			{
				SSLContext c;
				try (var x = Net.class.getResourceAsStream("localhost")) {
					c = Net.getSSLContext(Map.entry("JKS", x), "passphrase".toCharArray());
				}
				var p = Integer.parseInt(a.configuration.getProperty("todomvc.server.port"));
				s = a.diFactory.create(HttpServer.class,
						Map.of("sslContext", c, "endpoint", new InetSocketAddress(p), "handler", a.handler));
			}
			s.serve();
		} catch (Throwable e) {
			e.printStackTrace();
		}
	}

	protected final Properties configuration;

	protected final DiFactory diFactory;

	protected final List<Path> files;

	protected final HttpHandler handler;

	protected final List<Invocable> invocables;

	protected final RenderableFactory renderableFactory;

	public TodoMvc(DiFactory diFactory, Path configurationFile) {
		this.diFactory = diFactory;
		if (!INSTANCE.compareAndSet(null, this))
			throw new IllegalStateException();
		configuration = diFactory.create(Properties.class, Collections.singletonMap("file", configurationFile));

		invocables = types().stream()
				.flatMap(x -> Arrays.stream(x.getMethods())
						.filter(y -> !Modifier.isStatic(y.getModifiers()) && !y.isBridge())
						.map(y -> new Invocable(x, y)))
				.toList();
		files = Stream.of("com.janilla.frontend", TodoMvc.class.getPackageName())
				.flatMap(x -> Java.getPackagePaths(x).stream().filter(Files::isRegularFile)).toList();
		renderableFactory = diFactory.create(RenderableFactory.class);
		{
			var f = diFactory.create(ApplicationHandlerFactory.class);
			handler = x -> {
				var h = f.createHandler(Objects.requireNonNullElse(x.exception(), x.request()));
				if (h == null)
					throw new NotFoundException(x.request().getMethod() + " " + x.request().getTarget());
				return h.handle(x);
			};
		}
	}

	@Handle(method = "GET", path = "/")
	public TodoMvc application() {
		return this;
	}

	public Properties configuration() {
		return configuration;
	}

	public DiFactory diFactory() {
		return diFactory;
	}

	public List<Path> files() {
		return files;
	}

	public HttpHandler handler() {
		return handler;
	}

	public List<Invocable> invocables() {
		return invocables;
	}

	public RenderableFactory renderableFactory() {
		return renderableFactory;
	}

	public Collection<Class<?>> types() {
		return diFactory.types();
	}
}
